### PR TITLE
add promissory notes to TaskToCompute and SubtaskResultsVerify

### PR DESCRIPTION
**Pull request by shadeofblue**
closes: https://github.com/golemfactory/golem-messages/issues/328
closes: https://github.com/golemfactory/golem-messages/issues/329
closes: https://github.com/golemfactory/golem-messages/issues/330
closes: https://github.com/golemfactory/golem-messages/issues/331

### CODE REVIEW

File: golem_messages/datastructures/promissory.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    'PromissoryNoteSig',
+    [
+        ('v', int),
+        ('r', bytes),
+        ('s', bytes),
+    ]
+)
+
+
+class PromissoryNote:
```
Comment (jiivan) [inline golem_messages/datastructures/promissory.py]: Maybe it's a case of `datastructures.Container`?

File: golem_messages/datastructures/promissory.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        signed_message = eth_account.account.Account.signHash(  # noqa pylint: disable=no-value-for-parameter, line-too-long;  # pylint can't make sense of eth_utils' `@combomethod` decorator
+            message_hash=self.hash,
+            private_key=privkey
+        )
+        v = signed_message['v']
+        r = (signed_message['r']).to_bytes(32, byteorder='big')
+        s = (signed_message['s']).to_bytes(32, byteorder='big')
+        return PromissoryNoteSig(v, r, s)
+
+    def sig_valid(self, promissory_note_sig: typing.Optional[tuple]) -> bool:
```
Comment (jiivan) [inline golem_messages/datastructures/promissory.py]:
```suggestion
    def verify_signature(self, promissory_note_sig: typing.Optional[tuple]) -> bool:
```
As in `Message`

File: golem_messages/datastructures/promissory.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        if not promissory_note_sig:
+            return False
+
+        try:
+            promissory_note_sig = PromissoryNoteSig(*promissory_note_sig)
+        except TypeError as e:
+            raise TypeError(
```
Comment (jiivan) [inline golem_messages/datastructures/promissory.py]: This won't be needed if inherited from `Container`

File: golem_messages/utils.py [hunk 4]
```diff
@@ -2,9 +2,21 @@
 import copy
 from uuid import UUID
 
+from eth_utils import to_checksum_address
+from ethereum.utils import sha3
+
 from golem_messages import message
 
 
+def pubkey_to_address(pubkey: str) -> str:
```
Comment (tworec) [inline golem_messages/utils.py]: there at least two available implementations of this available:
* in bitcoin that is already a dep for GM [src](https://github.com/vbuterin/pybitcointools/blob/aeb0a2bbb8bbfe421432d776c649650eaeb882a5/bitcoin/main.py#L450)
* in pyethereum [src](https://github.com/ethereum/pyethereum/blob/b704a5c6577863edc539a1ec3d2620a443b950fb/ethereum/utils.py#L119)

maybe you can use it...

Comment (Krigpl) [reply golem_messages/utils.py]: Bitcoin implementation returns a Bitcoin address - not what we want
this pyethereum implementation that you linked returns an address of a contract that will be created, not something that applies here
Last time I checked there was no implementation of this kind, which is weird, but that's the case.

File: golem_messages/datastructures/promissory.py (hunk 1 above)
Comment (shadeofblue) [reply golem_messages/datastructures/promissory.py]: huh? it doesn't need to be serialized/deserialized...

File: golem_messages/datastructures/promissory.py (hunk 2 above)
Comment (shadeofblue) [reply golem_messages/datastructures/promissory.py]: meh, it's a different usage than in `Message` ... it doesn't need to be similar imvho

File: golem_messages/datastructures/promissory.py (hunk 3 above)
Comment (shadeofblue) [reply golem_messages/datastructures/promissory.py]: are we serious that we want to inherit a _tuple_ from a more complicated class? ... I'd rather leave it as is

File: golem_messages/utils.py [hunk 5]
```diff
@@ -2,9 +2,21 @@
 import copy
 from uuid import UUID
 
+from eth_utils import to_checksum_address
+from ethereum.utils import sha3
```
[the remark this answers is not in the document]
Comment (shadeofblue) [reply golem_messages/utils.py]: done

File: golem_messages/datastructures/promissory.py (hunk 3 above)
Comment (jiivan) [reply golem_messages/datastructures/promissory.py]: I mistakenly thought that it's about `PromissoryNote`. Sorry.

**Pull request merged**